File: python/memory_test.py
#!/usr/bin/env python3
import psutil
import os
import time
import sys
import gc

def memory_usage():
    """获取当前进程的内存使用量（MB）"""
    process = psutil.Process(os.getpid())
    return process.memory_info().rss / (1024 * 1024)

def run_memory_test(array_size):
    """测试创建和操作大数组的内存使用情况"""
    # 强制垃圾回收
    gc.collect()
    time.sleep(0.1)
    
    # 初始内存使用
    initial_memory = memory_usage()
    print(f"初始内存使用: {initial_memory:.2f} MB")
    
    # 创建大数组
    print(f"创建大小为 {array_size} 的数组...")
    start_time = time.time()
    large_array = [i for i in range(array_size)]
    creation_time = time.time() - start_time
    
    after_creation_memory = memory_usage()
    print(f"数组创建后内存使用: {after_creation_memory:.2f} MB (增加 {after_creation_memory - initial_memory:.2f} MB)")
    print(f"数组创建时间: {creation_time:.6f} 秒")
    
    # 对数组进行操作
    print("执行数组求和操作...")
    start_time = time.time()
    sum_result = sum(large_array)
    operation_time = time.time() - start_time
    
    after_operation_memory = memory_usage()
    print(f"操作后内存使用: {after_operation_memory:.2f} MB (增加 {after_operation_memory - after_creation_memory:.2f} MB)")
    print(f"数组求和结果: {sum_result}, 操作时间: {operation_time:.6f} 秒")
    
    # 清理
    del large_array
    gc.collect()
    time.sleep(0.1)
    
    final_memory = memory_usage()
    print(f"清理后内存使用: {final_memory:.2f} MB")

if __name__ == "__main__":
    array_size = 10000000  # 1千万元素
    if len(sys.argv) > 1:
        array_size = int(sys.argv[1])
    
    run_memory_test(array_size)

File: nodejs/memory_test.js
/**
 * 获取当前进程内存使用情况（MB）
 */
function getMemoryUsage() {
    const memoryUsage = process.memoryUsage();
    return {
        rss: memoryUsage.rss / (1024 * 1024),
        heapTotal: memoryUsage.heapTotal / (1024 * 1024),
        heapUsed: memoryUsage.heapUsed / (1024 * 1024),
        external: memoryUsage.external / (1024 * 1024)
    };
}

/**
 * 内存使用测试
 */
async function runMemoryTest(arraySize) {
    // 强制垃圾回收（如果可能的话）
    if (global.gc) {
        global.gc();
    } else {
        console.log("提示: 运行时带上 --expose-gc 标志以启用垃圾回收");
    }
    
    // 等待100ms让可能的GC完成
    await new Promise(resolve => setTimeout(resolve, 100));
    
    // 初始内存使用
    const initialMemory = getMemoryUsage();
    console.log(`初始内存使用: RSS = ${initialMemory.rss.toFixed(2)} MB, 堆使用 = ${initialMemory.heapUsed.toFixed(2)} MB`);
    
    // 创建大数组
    console.log(`创建大小为 ${arraySize} 的数组...`);
    const startTime = process.hrtime.bigint();
    const largeArray = Array(arraySize).fill().map((_, i) => i);
    const creationTime = Number(process.hrtime.bigint() - startTime) / 1e9;
    
    const afterCreationMemory = getMemoryUsage();
    console.log(`数组创建后内存使用: RSS = ${afterCreationMemory.rss.toFixed(2)} MB (增加 ${(afterCreationMemory.rss - initialMemory.rss).toFixed(2)} MB)`);
    console.log(`堆使用 = ${afterCreationMemory.heapUsed.toFixed(2)} MB (增加 ${(afterCreationMemory.heapUsed - initialMemory.heapUsed).toFixed(2)} MB)`);
    console.log(`数组创建时间: ${creationTime.toFixed(6)} 秒`);
    
    // 对数组进行操作
    console.log("执行数组求和操作...");
    const startOpTime = process.hrtime.bigint();
    const sumResult = largeArray.reduce((acc, val) => acc + val, 0);
    const operationTime = Number(process.hrtime.bigint() - startOpTime) / 1e9;
    
    const afterOperationMemory = getMemoryUsage();
    console.log(`操作后内存使用: RSS = ${afterOperationMemory.rss.toFixed(2)} MB (增加 ${(afterOperationMemory.rss - afterCreationMemory.rss).toFixed(2)} MB)`);
    console.log(`堆使用 = ${afterOperationMemory.heapUsed.toFixed(2)} MB (增加 ${(afterOperationMemory.heapUsed - afterCreationMemory.heapUsed).toFixed(2)} MB)`);
    console.log(`数组求和结果: ${sumResult}, 操作时间: ${operationTime.toFixed(6)} 秒`);
    
    // 清理
    largeArray.length = 0;
    if (global.gc) {
        global.gc();
        await new Promise(resolve => setTimeout(resolve, 100));
    }
    
    const finalMemory = getMemoryUsage();
    console.log(`清理后内存使用: RSS = ${finalMemory.rss.toFixed(2)} MB, 堆使用 = ${finalMemory.heapUsed.toFixed(2)} MB`);
}

async function main() {
    // 从命令行获取参数或使用默认值
    const arraySize = process.argv.length > 2 ? parseInt(process.argv[2]) : 10000000;
    
    await runMemoryTest(arraySize);
}

main().catch(console.error);